### PR TITLE
[inductor] add cpp builder code. (take 2)

### PR DESCRIPTION
**Pull request by xuhancn**
Fully manual rebase the code of PR: https://github.com/pytorch/pytorch/pull/124045
The old PR seems crashed due to too many commits, and too many times rebase. Please reference: https://github.com/pytorch/pytorch/pull/124045#issuecomment-2103744588

-------
It is the first step of RFC https://github.com/pytorch/pytorch/issues/124245.
Changes:
1. Add cpp builder code, the new cpp_builder support Windows OS.
2. Add CPU ISA checker which is cross OS and exported from backend cpuinfo.
3. Switch compiler ISA checker to new cpp builder.
4. CppCodeCache use the new ISA checker.
5. Add temprary `test_new_cpp_build_logical` UT to help on transfer to new code.
Image

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Comment by jansel** [comment]
I'll let @desertfire handle the review/landing here.

**Comment by desertfire** [comment]
Internal test failure, `CUDA_HOME` not set, although I haven't dug into why.
```
  File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/codecache.py", line 2320, in load_async
    name="o", sources="i", BuildOption=CppTorchCudaOptions(**compile_command)
  File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/cpp_builder.py", line 912, in __init__
    ) = get_cpp_torch_cuda_options(aot_mode=aot_mode)
  File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/cpp_builder.py", line 855, in get_cpp_torch_cuda_options
    ) = _get_cuda_related_args(aot_mode)
  File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/cpp_builder.py", line 779, in _get_cuda_related_args
    include_dirs = cpp_extension.include_paths(cuda)
  File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/utils/cpp_extension.py", line 1171, in include_paths
    cuda_home_include = _join_cuda_home('include')
  File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/utils/cpp_extension.py", line 2417, in _join_cuda_home
    raise OSError('CUDA_HOME environment variable is not set. '
 ```

**Comment by xuhancn** [comment]
> Internal test failure, `CUDA_HOME` not set, although I haven't dug into why.
> 
> ```
>   File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/codecache.py", line 2320, in load_async
>     name="o", sources="i", BuildOption=CppTorchCudaOptions(**compile_command)
>   File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/cpp_builder.py", line 912, in __init__
>     ) = get_cpp_torch_cuda_options(aot_mode=aot_mode)
>   File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/cpp_builder.py", line 855, in get_cpp_torch_cuda_options
>     ) = _get_cuda_related_args(aot_mode)
>   File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/_inductor/cpp_builder.py", line 779, in _get_cuda_related_args
>     include_dirs = cpp_extension.include_paths(cuda)
>   File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/utils/cpp_extension.py", line 1171, in include_paths
>     cuda_home_include = _join_cuda_home('include')
>   File "/dev/shm/uid-30083/5b4bda56-seed-nspid4026534425_cgpid3897394-ns-4026534632/torch/utils/cpp_extension.py", line 2417, in _join_cuda_home
>     raise OSError('CUDA_HOME environment variable is not set. '
> ```

Some debug discussion sent to you via slack, thanks. @desertfire

**Comment by desertfire** [comment]
> Some debug discussion sent to you via slack, thanks. @desertfire

https://github.com/pytorch/pytorch/pull/126408/ should solve this CUDA_HOME not set issue.

**Comment by int3** [comment]
JFYI I'm on PTO and won't be able to land https://github.com/pytorch/pytorch/pull/126408 till Tuesday

**Comment by xuhancn** [comment]
> > Some debug discussion sent to you via slack, thanks. @desertfire
> 
> #126408 should solve this CUDA_HOME not set issue. You need to rebase after that one lands. After that, I will give another try in fbcode.

I have synced code to https://github.com/pytorch/pytorch/pull/126408 , @desertfire please help on import fb_code to validate, much appreciated.

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -2322,8 +2373,19 @@ def load_async(cls, source_code: str, cuda=False, submit_fn=None, extra_flags=()
             "vec_isa": pick_vec_isa(),
             "extra_flags": extra_flags,
         }
-        cpp_command = repr(cpp_compile_command("i", "o", **compile_command))
-        key, input_path = write(source_code, "cpp", extra=cpp_command)
+
```
Comment (desertfire) [inline torch/_inductor/codecache.py:2376]: Thanks to @int3 , adding `_set_gpu_runtime_env()  # cpp_extension consults the env` here makes fbcode tests pass.

**Review by desertfire — CHANGES_REQUESTED** [review]

**Comment by xuhancn** [comment]
@desertfire apply code follow your request, please take a look.

**Comment by xuhancn** [comment]
@desertfire fixed some fb_code issues from new cpp_builder.

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -2620,6 +2684,99 @@ class CppWrapperCodeCache(CppPythonBindingsCodeCache):
...
+def _do_validate_cpp_commands(
+    include_pytorch: bool,
+    cuda: bool,
+    compile_only: bool,
+    mmap_weights: bool,
+    use_absolute_path: bool,
+):
+    # PreCI will failed if test machine can't run cuda.
+    test_cuda = torch.cuda.is_available() and cuda
+    input_path = "/temp/dummy_input.cpp"
```
Comment (desertfire) [inline torch/_inductor/codecache.py:2709]: Can you use `tempfile.TemporaryDirectory()` instead of hard-code it?

**Comment by desertfire** [comment]
test_new_cpp_build_logical still fails because of missing openmp and python include paths, https://github.com/pytorch/pytorch/blob/30788739f4cec672cdabc21c155bbe30ac5d66e6/torch/_inductor/codecache.py#L1755-L1756 . Other than that, there are macro definition differences like '-DC10_USE_GLOG' vs. '-D C10_USE_GLOG' which is benign, but it does mean you need to either relax your testing, or just skip it for fbcode.

**Comment by xuhancn** [comment]
@desertfire try to fixed use_absolute_path in fb_code.

**Comment by xuhancn** [comment]
@desertfire fix fb_code issues:
1. omp lib.
2. torch, torch_cpu and torch_python libs.

**Comment by xuhancn** [comment]
@desertfire update code:
1. Sync _get_python_include_dirs to new cpp_builder.
2. Fix fb_code cuda include.

**Review by desertfire — APPROVED** [review]

**Comment by xuhancn** [comment]
> ## Merge failed
> **Reason**: 1 jobs have failed, first few of them are: [Meta Internal-Only Changes Check](https://opensource.facebook.com/)
> 
> Details for Dev Infra team

@desertfire How to solve this issue?

**Pull request closed without merging**